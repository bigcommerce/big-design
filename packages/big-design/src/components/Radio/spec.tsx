import { fireEvent, render, screen } from '@testing-library/react';
import React, { createRef } from 'react';
import 'jest-styled-components';

import { warning } from '../../utils';

import { Radio, RadioLabel } from './index';

test('render Radio (checked)', () => {
  render(<Radio checked={true} label="Checked" name="test-group" onChange={() => null} />);

  expect(screen.getByRole('radio')).toMatchSnapshot();
});

test('render Radio (unchecked)', () => {
  render(<Radio checked={false} label="Unchecked" name="test-group" onChange={() => null} />);

  expect(screen.getByRole('radio')).toMatchSnapshot();
});

test('render Radio (unchecked + description object)', () => {
  render(
    <Radio
      checked={false}
      description={{
        text: 'description',
        link: {
          text: 'learn more',
          target: 'foo',
          href: 'bar',
        },
      }}
      label="Unchecked"
      name="test-group"
      onChange={() => null}
    />,
  );

  expect(screen.getByRole('radio')).toMatchSnapshot();
});

test('render Radio (unchecked + description text)', () => {
  render(
    <Radio
      checked={false}
      description="description"
      label="Unchecked"
      name="test-group"
      onChange={() => null}
    />,
  );

  expect(screen.getByRole('radio')).toMatchSnapshot();
});

test('render Radio (checked + disabled)', () => {
  render(<Radio checked={true} disabled label="Checked" name="test-group" onChange={() => null} />);

  expect(screen.getByRole('radio')).toMatchSnapshot();
});

test('render Radio (unchecked + disabled)', () => {
  render(
    <Radio checked={false} disabled label="Unchecked" name="test-group" onChange={() => null} />,
  );

  expect(screen.getByRole('radio')).toMatchSnapshot();
});

test('has correct value when checked', async () => {
  render(<Radio checked={true} data-testid="radio" label="Checked" onChange={() => null} />);

  const radio = await screen.findByTestId<HTMLInputElement>('radio');

  expect(radio).toBeChecked();
});

test('has correct value when unchecked', async () => {
  render(<Radio checked={false} data-testid="radio" label="Unchecked" onChange={() => null} />);

  const radio = await screen.findByTestId<HTMLInputElement>('radio');

  expect(radio).not.toBeChecked();
});

test('triggers onChange when clicking the radio button', async () => {
  const onChange = jest.fn();

  render(<Radio checked={false} data-testid="radio" label="Checked" onChange={onChange} />);

  const radio = await screen.findByTestId<HTMLInputElement>('radio');

  fireEvent.click(radio);

  expect(onChange).toHaveBeenCalled();
});

test('triggers onChange when clicking styled and text label', () => {
  const onChange = jest.fn();

  const { container } = render(<Radio checked={false} label="Checked" onChange={onChange} />);

  // Click the styled radio (visual element)
  // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
  const styledRadio = container.querySelector('label[aria-hidden="true"]');

  if (styledRadio) {
    fireEvent.click(styledRadio);
  }

  // Click the text label
  const textLabel = screen.getByText('Checked');

  fireEvent.click(textLabel);

  expect(onChange).toHaveBeenCalledTimes(2);
});

test('accepts valid RadioLabel component', () => {
  const testId = 'test';
  const label = <RadioLabel data-testid={testId}>Label</RadioLabel>;

  render(<Radio label={label} />);

  expect(screen.getByTestId(testId)).toBeInTheDocument();
});

test('does not accept invalid label component', () => {
  const testId = 'test';
  const label = <div data-testid={testId}>Label</div>;

  render(<Radio label={label} />);

  expect(warning).toHaveBeenCalledTimes(1);
  expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
});

test('forwards ref', () => {
  const ref = createRef<HTMLInputElement>();

  render(<Radio data-testid="radio" label="Checked" ref={ref} />);

  expect(screen.getByTestId('radio')).toBe(ref.current);
});

test('does not forward styles', () => {
  render(<Radio className="test" label="Checked" />);

  expect(screen.queryByRole('radio', { name: /test/i })).not.toBeInTheDocument();
});
