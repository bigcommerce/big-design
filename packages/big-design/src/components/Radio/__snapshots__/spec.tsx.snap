// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`render Radio (checked + disabled) 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-labelledby=":r9:"
  checked=""
  class="c0"
  disabled=""
  id=":r8:"
  name="test-group"
  type="radio"
/>
`;

exports[`render Radio (checked) 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-labelledby=":r1:"
  checked=""
  class="c0"
  id=":r0:"
  name="test-group"
  type="radio"
/>
`;

exports[`render Radio (unchecked + description object) 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-labelledby=":r5:"
  class="c0"
  id=":r4:"
  name="test-group"
  type="radio"
/>
`;

exports[`render Radio (unchecked + description text) 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-labelledby=":r7:"
  class="c0"
  id=":r6:"
  name="test-group"
  type="radio"
/>
`;

exports[`render Radio (unchecked + disabled) 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-labelledby=":rb:"
  class="c0"
  disabled=""
  id=":ra:"
  name="test-group"
  type="radio"
/>
`;

exports[`render Radio (unchecked) 1`] = `
.c0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<input
  aria-labelledby=":r3:"
  class="c0"
  id=":r2:"
  name="test-group"
  type="radio"
/>
`;
